cmd/site-replication-metrics.go

```diff
@@ -0,0 +1,285 @@
+	// Total number of completed operations
+	ReplicatedCount int64 `json:"replicationCount"`
+	// Replication latency information
+	Latency ReplicationLatency `json:"replicationLatency"`
+	// transfer rate for large uploads
+	XferRateLrg *XferStats `json:"largeTransferRate" msg:"lt"`
+	// transfer rate for small uploads
+	XferRateSml *XferStats `json:"smallTransferRate" msg:"st"`
+	// Endpoint is the replication target endpoint
+	Endpoint string `json:"-"`
+	// Secure is true if the replication target endpoint is secure
+	Secure bool `json:"-"`
+}
+
+func (sr *SRStats) update(st replStat, dID string) {
+	sr.lock.Lock()
+	defer sr.lock.Unlock()
+	srs, ok := sr.M[dID]
+	if !ok {
+		srs = &SRStatus{
+			XferRateLrg: newXferStats(),
+			XferRateSml: newXferStats(),
+		}
+	}
+	srs.Endpoint = st.Endpoint
+	srs.Secure = st.Secure
+	switch {
+	case st.Completed:
+		srs.ReplicatedSize += st.TransferSize
+		srs.ReplicatedCount++
+		if st.TransferDuration > 0 {
+			srs.Latency.update(st.TransferSize, st.TransferDuration)
+			srs.updateXferRate(st.TransferSize, st.TransferDuration)
+		}
+	case st.Failed:
+		srs.Failed.addsize(st.TransferSize, st.Err)
+	case st.Pending:
+	}
+	sr.M[dID] = srs
+}
+
+func (sr *SRStats) get() map[string]SRMetric {
+	epMap := globalBucketTargetSys.healthStats()
+
+	sr.lock.RLock()
+	defer sr.lock.RUnlock()
+	m := make(map[string]SRMetric, len(sr.M))
+	for dID, v := range sr.M {
+		t := newXferStats()
+		mx := make(map[RMetricName]XferStats)
+
+		if v.XferRateLrg != nil {
+			mx[Large] = *v.XferRateLrg.Clone()
+			m := t.merge(*v.XferRateLrg)
+			t = &m
+		}
+		if v.XferRateSml != nil {
+			mx[Small] = *v.XferRateSml.Clone()
+			m := t.merge(*v.XferRateSml)
+			t = &m
+		}
+
+		mx[Total] = *t
+		metric := SRMetric{
+			ReplicatedSize:  v.ReplicatedSize,
+			ReplicatedCount: v.ReplicatedCount,
+			DeploymentID:    dID,
+			Failed:          v.Failed.toMetric(),
+			XferStats:       mx,
+		}
+		epHealth, ok := epMap[v.Endpoint]
+		if ok {
+			metric.Endpoint = epHealth.Endpoint
+			metric.TotalDowntime = epHealth.offlineDuration
+			metric.LastOnline = epHealth.lastOnline
+			metric.Online = epHealth.Online
+			metric.Latency = madmin.LatencyStat{
+				Curr: epHealth.latency.curr,
+				Avg:  epHealth.latency.avg,
+				Max:  epHealth.latency.peak,
+			}
+		}
+		m[dID] = metric
+	}
+	return m
+}
+
+func (srs *SRStatus) updateXferRate(sz int64, duration time.Duration) {
+	if sz > minLargeObjSize {
+		srs.XferRateLrg.addSize(sz, duration)
+	} else {
+		srs.XferRateSml.addSize(sz, duration)
+	}
+}
+
+func newSRStats() *SRStats {
+	s := SRStats{
+		M:               make(map[string]*SRStatus),
+		movingAvgTicker: time.NewTicker(time.Second * 2),
+	}
+	go s.trackEWMA()
+	return &s
+}
+
+func (sr *SRStats) trackEWMA() {
+	for {
+		select {
+		case <-sr.movingAvgTicker.C:
+			sr.updateMovingAvg()
+		case <-GlobalContext.Done():
+			return
+		}
+	}
+}
+
+func (sr *SRStats) updateMovingAvg() {
+	sr.lock.Lock()
+	defer sr.lock.Unlock()
+}
```

Comment:
this function is not doing anything - did you miss implementing this?

Reply:
it was being done along with bucket replication updates - moved it into above function